osu.Game/Overlays/Practice/PracticePlayer.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Beatmaps;
+using osu.Game.Graphics;
+using osu.Game.Screens.Play;
+
+namespace osu.Game.Overlays.Practice
+{
+    public class PracticePlayer : Player
+    {
+        private PracticeOverlay practiceOverlay = null!;
+
+        public PracticePlayer(PlayerConfiguration? configuration = null)
+            : base(configuration)
+        {
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colour, IBindable<WorkingBeatmap> beatmap, PracticePlayerLoader loader)
+        {
+            var playableBeatmap = beatmap.Value.GetPlayableBeatmap(beatmap.Value.BeatmapInfo.Ruleset);
+
+            SetGameplayStartTime(loader.CustomStart.Value * (playableBeatmap.HitObjects.Last().StartTime - playableBeatmap.HitObjects.First().StartTime));
```

Comment:
why is this done using DI? the loader creates the player so you can call directly and avoid the copmlexity.

Reply:
sorry for the late response, was attempting to fix it somehow but im out of ideas, to put it simply i think this is the only case where the the playerloader ( well, practiceplayerloader here ) needs to be able to pass data to the player being created which doesnt exist until after the playerloader itself has been created ( except for ofcourse the base playerloader class which has access to more stuff) so i opted to DI the playerloader into the player after creation to feed it the data, instead of altering the base playerloader to allow its derived classes to create the player from within themselves.